implement addNodeAddonIncludePaths

bmacnaughton opened the pull request:
This PR addresses the common cases for #4854 by adding the include paths specified by `nan` and `node-addon-api` when they are listed as dependencies in package.json. This was the first time I've looked at vscode-cpptools code and it's also the first time I've written any typescript, so suggestions about how I can improve in either area are welcome.

I haven't changed the changelog nor even looked at adding a specific test pending feedback. Afaict, it passes the unit tests and integration tests but I don't see any positive feedback to be sure.

I modified the insiders branch, so that is the target for this PR. Not sure whether that's the regular practice or not.

Questions:

- Is "machine-overridable" the correct scope [package.json 1109](https://github.com/microsoft/vscode-cpptools/compare/insiders...bmacnaughton:node-addon-includes?expand=1#diff-62bdf835c36931e2307d4d55bb2f11bcfef5e8dc5714ef35b7d5ae81bc5a10bcR1109)
- should it check `settings.addNodeAddonIncludePaths` or just concat an empty array [configuration.ts 327](https://github.com/microsoft/vscode-cpptools/compare/insiders...bmacnaughton:node-addon-includes?expand=1#diff-9b32cba7ca4094f68784eb207728f8c4dd5aa6f0a3aa601c01c55fec735b589eR327)
- mapping the dependencies to commands uses an inline definition. not sure the best practices for this [configuration.ts 398](https://github.com/microsoft/vscode-cpptools/compare/insiders...bmacnaughton:node-addon-includes?expand=1#diff-9b32cba7ca4094f68784eb207728f8c4dd5aa6f0a3aa601c01c55fec735b589eR327).
- `pfs` (fs/promises) is defined at the top of the file but only used one place [configuration.ts 408](https://github.com/microsoft/vscode-cpptools/compare/insiders...bmacnaughton:node-addon-includes?expand=1#diff-9b32cba7ca4094f68784eb207728f8c4dd5aa6f0a3aa601c01c55fec735b589eR327). prefer it locally scoped?
- the include path loop uses a labeled loop. I didn't see those anywhere else; it could be replaced with an if-level. [configuration.ts 414](https://github.com/microsoft/vscode-cpptools/compare/insiders...bmacnaughton:node-addon-includes?expand=1#diff-9b32cba7ca4094f68784eb207728f8c4dd5aa6f0a3aa601c01c55fec735b589eR327)

Obviously, you're welcome to change names, wording, etc. or ask me to do so.

bmacnaughton committed:
add addNodeAddonIncludePaths

sean-mcmanus reviewed (changes requested):
You should install the ESLint extension, and run the the "lint" npm script or "gulp lint". The linter enforces are our whitespace/style rules.

configurations.ts
   24:7   error  expected variable-declaration: 'pfs' to have a typedef (tslint:typedef)           @typescript-eslint/tslint/config
  399:1   error  Expected indentation of 8 spaces but found 6                                      @typescript-eslint/indent
  400:1   error  Expected indentation of 8 spaces but found 6                                      @typescript-eslint/indent
  400:86  error  Unexpected trailing comma                                                         comma-dangle
  403:66  error  Unexpected space before the ':'                                                   @typescript-eslint/type-annotation-spacing       
  408:23  error  expected variable-declaration: 'package_json' to have a typedef (tslint:typedef)  @typescript-eslint/tslint/config
  411:1   error  Expected indentation of 20 spaces but found 18                                    @typescript-eslint/indent
  412:1   error  Expected indentation of 24 spaces but found 22                                    @typescript-eslint/indent
  413:1   error  Expected indentation of 24 spaces but found 22                                    @typescript-eslint/indent
  413:27  error  expected variable-declaration: 'stdout' to have a typedef (tslint:typedef)        @typescript-eslint/tslint/config
  414:1   error  Expected indentation of 24 spaces but found 22                                    @typescript-eslint/indent
  414:23  error  Expected { after 'if' condition                                                   curly
  416:1   error  Expected indentation of 24 spaces but found 22                                    @typescript-eslint/indent
  417:1   error  Expected indentation of 24 spaces but found 22                                    @typescript-eslint/indent
  418:1   error  Expected indentation of 28 spaces but found 26                                    @typescript-eslint/indent
  419:1   error  Expected indentation of 24 spaces but found 22                                    @typescript-eslint/indent
  420:1   error  Expected indentation of 24 spaces but found 22                                    @typescript-eslint/indent
  421:1   error  Expected indentation of 24 spaces but found 22                                    @typescript-eslint/indent
  422:1   error  Expected indentation of 28 spaces but found 26                                    @typescript-eslint/indent
  423:1   error  Expected indentation of 24 spaces but found 22                                    @typescript-eslint/indent
  424:1   error  Expected indentation of 24 spaces but found 22                                    @typescript-eslint/indent
  425:1   error  Expected indentation of 28 spaces but found 26                                    @typescript-eslint/indent
  426:1   error  Expected indentation of 24 spaces but found 22                                    @typescript-eslint/indent
  431:1   error  Expected indentation of 12 spaces but found 10                                    @typescript-eslint/indent

✖ 24 problems (24 errors, 0 warnings)
  21 errors and 0 warnings potentially fixable with the `--fix` option.

sean-mcmanus commented on Extension/src/LanguageServer/configurations.ts:
```diff
@@ -317,6 +323,9 @@ export class CppProperties {
         } else {
             configuration.includePath = [defaultFolder];
         }
+        // check settings.addNodeAddonIncludePaths?
```
Yeah, can you check settings.addNodeAddonIncludePaths and skip the concat if it's false? -- that would enable dynamic disabling without a reload. For the enabling case, in onDidChangeSettings you can just check for the addNodeAddonIncludePaths settings being enabled and then call promptForReloadWindowDueToSettingsChange ...assuming you don't want to make dynamic enabling possible.

bmacnaughton replied:
i used `isUnset()` because all the other checks do but am wondering why `!settings.addNodeAddonIncludePaths` might not be valid.

i'm now looking at *which* `onDidChangeSettings()` would be appropriate to modify. any hints appreciated as i don't see any of them that (directly) call `util.promptForReloadWindowDueToSettingsChange()`.

sean-mcmanus replied:
IsUnset is used because in TypeScript values like "" also implicitly convert to false in "if" statements. So using `isUnset` is not what you want, since it just checks if the setting doesn't exist.

Looking more closely at the code, `applyDefaultConfigurationValues` maybe be the wrong place for setting this. The reason the vcpkgIncludes is set here is because we wanted it to only appear when users create a new default c_cpp_properties.json, but in your case it sounds like you want it to always be auto-added?

The correct spot may be `updateServerOnFolderSettingsChange` (I'm not 100% sure).

Yeah, we currently call util.promtForReloadWindowDueToSettingsChange after we get a response from cpptools from the setting change. I was referring to `DefaultClient.onDidChangeSettings(event: vscode.ConfigurationChangeEvent, isFirstClient: boolean): { [key: string]: string }`, which has some existing code that checks if certain settings were changed.

bmacnaughton replied:
> but in your case it sounds like you want it to always be auto-added

correct. those paths should always be auto-added. this is where i am working now. it doesn't appear that `updateServerOnFolderSettingsChange()` would work if the function `addNodeAddonIncludeLocations` always reads package.json and sets `configuration.nodeAddonIncludes`. right now that is only set at startup when the setting 'addNodeAddonIncludePaths` is set, so removing that check results in those not being read at startup. but it looks like i might need a call to `addNodeAddonIncludeLocations` in both places to handle startup and a settings change.

i'll keep debugging and trying to determine how to get the logic working. i ran into the "unbound breakpoint" problem but was able to recover by checking out a previous commit. it took me a bit though.

sean-mcmanus replied:
> * i'm thinking of renaming `addNodeAddonIncludePaths` to `readNodeAddonIncludePaths` because it doesn't add them.

That sounds good.

> (mostly allows changing on the fly) if the user adds `nan` or `node-addon-api` to the project on the fly the code doesn't know that and doesn't know to re-execute `addNodeAddonIncludePaths()`. is there a way to get notified when a specific file (the project's package.json) is modified?

Yes, you can add a file watcher on the package.json and react to changes. We do the same thing for c_cpp_properties.json at https://github.com/microsoft/vscode-cpptools/blob/insiders/Extension/src/LanguageServer/configurations.ts#L207 .

In some cases, the file watchers may file, so you could fallback to checking the time stamp in our idle loop processing like was done in https://github.com/microsoft/vscode-cpptools/pull/6748/files for compile_commands.json .
 
> and i need to clean up the `showErrorMessage('error.adding.node.addon.includes')` but wanted to get your feedback on the preferred way to showing errors like this first.
> 
> feedback is appreciated.

I didn't get a chance to finish reviewing the changes today.

sean-mcmanus commented on Extension/src/LanguageServer/configurations.ts:
```diff
@@ -386,6 +395,46 @@ export class CppProperties {
+                          stdout = stdout.slice(1, -1);
+                      }
+                      if (stdout) {
+                          this.nodeAddonIncludes.push(stdout);
```
I think you should check if the stdout is a valid directory, maybe calling util.checkDirectoryExists. If it doesn't exist, you could check if path.join("${workspaceFolder}", stdout) exists, since I'm getting back a relative path `node_modules\\nan` from nan.

bmacnaughton replied:
nan always returns the path relative to the working directory, so i'm calling `path.resolve()` on the results and checking that. i think the only time `util.checkDirectoryExists()` will fail is if the node command executed by `addNodeAddonIncludeLocations()` fails.

sean-mcmanus replied:
path.resolve doesn't work because it uses a working path of VS Code, such as `C:\\Users\\<user>\\AppData\\Local\\Programs\\Microsoft VS Code\`, and not the user's workspace. Was there a reason `path.join("${workspaceFolder}", stdout)` or something similar wouldn't work?

bmacnaughton replied:
no reason that doesn't work. i was trying to get it resolved with one shot as opposed to looking around but obviously hadn't checked my assumption about the working directory. i will address this in the next push.

sean-mcmanus commented on Extension/src/LanguageServer/configurations.ts:
```diff
@@ -386,6 +395,46 @@ export class CppProperties {
+                  if (dep in package_json.dependencies) {
+                      const execCmd: string = this.nodeAddonMap[dep];
+                      let stdout = await util.execChildProcess(execCmd, rootPath);
+                      if (!stdout) continue dependencyLoop;
```
This can just be "continue", right? I don't see why the label is needed.

bmacnaughton replied:
i use labels so that if the code changes there is no need to worry about the scope of the continue. but i'm happy to change it.

sean-mcmanus replied:
Yeah, I think the label should be removed. It is expected that in the rare a case an sub-loop is added inside another loop then any "continue" would need to be addressed.

bmacnaughton committed:
partial review comments fixes

sean-mcmanus commented on Extension/package.json:
```diff
@@ -1102,6 +1102,12 @@
+          "type": "boolean",
+          "default": false,
+          "markdownDescription": "%c_cpp.configuration.addNodeAddonIncludePaths.description%",
+          "scope": "resource"
```
On second thought, I think this should be "application" scope (i.e. User scope in the UI), because otherwise it would be impossible for a user to download/open/browse a workspace from the internet securely, because execChildProcess is called. Also, it seems rare to want to be enable/disable the setting on a per-project basis anyway.

bmacnaughton replied:
ok. the scope definitions are still a bit of a mystery to me but next push will have this change.

sean-mcmanus commented on Extension/src/LanguageServer/configurations.ts:
```diff
@@ -386,6 +396,52 @@ export class CppProperties {
+
+            const settings: CppSettings = new CppSettings(this.rootUri);
+            if (settings.addNodeAddonIncludePaths) {
+                const package_json: any = JSON.parse(await pfs.readFile(path.join(rootPath, "package.json"), "utf8"));
```
It seems like it would be better to check for this file's existence first?

bmacnaughton replied:
yes, i should or the user will get a message about an unhandled promise rejection.

i will change this logic to better capture errors and issue better messages. should i use `vscode.window.showErrorMessage()` to display a message if there is a problem? it doesn't seem like using console.log is optimal.

sean-mcmanus replied:
For serious errors that users should always see, showErrorMessage should be used, the console.log may only get seen if the user looks at the logs or does a Toggle Developer Console. For this case, it seems like it should just silently skip the processing, since the user would just be opening a non-nodejs project.

bmacnaughton committed:
change scope to "application"

bmacnaughton committed:
make setting dynamic

- set at init and on changes
- always read nan/node-addon-api paths

bmacnaughton committed:
restructure addon include reads

bmacnaughton committed:
prompt for reload if no addon paths

- make nodeAddonMap a local
- fail silently
- add nodeAddonIncludesFound()

sean-mcmanus commented on Extension/src/LanguageServer/configurations.ts:
```diff
@@ -386,6 +395,71 @@ export class CppProperties {
+            "node-addon-api": "node --no-warnings -p \"require('node-addon-api').include\""
+        };
+
+        let added: number = 0;
```
This isn't being used currently. Should it be removed?

bmacnaughton replied:
yes, nice eye - thanks. it was an experiment to avoid `handleConfigurationChange()` but it didn't work correctly.

sean-mcmanus commented on Extension/src/LanguageServer/configurations.ts:
```diff
@@ -386,6 +395,71 @@ export class CppProperties {
+        if (error) {
+            // if it's possible for the directories returned by nan or node-addon-api to fail then
+            // this should be uncommented.
+            // console.log('readNodeAddonIncludeLocations', error.message);
```
The logging seems like it would be okay as long as the missing package.json case is excluded (currently it's not).

bmacnaughton replied:
done

sean-mcmanus commented on Extension/src/LanguageServer/configurations.ts:
```diff
@@ -317,6 +321,11 @@ export class CppProperties {
         }
+
+        if (settings.addNodeAddonIncludePaths) {
+            configuration.includePath = configuration.includePath.concat(this.nodeAddonIncludes);
```
It seems like the changes to applyDefaultIncludePathsAndFrameworks should be removed, because it sounds like you don't want the default c_cpp_properties.json to have the nodeAddonIncludes added...otherwise they would be added twice by updateServerOnFolderSettingsChange. The reason we added vcpkg paths to the default was because we expected that to be user modified.

bmacnaughton replied:
i can remove but if i do it creates this problem (which maybe i should be handling in some other way):

- start with "Add Node Addon Include Paths" checked - no squiggles
- uncheck "Add Node Addon Include Paths" - get squiggles
- check "Add Node Addon Include Paths" - squiggles do not go away

with the lines as they are the squiggles go away when the setting is checked (3rd bullet).

bmacnaughton replied:
i will figure it out. it's not clear why that makes a difference but it's not working the way i thought it was.

bmacnaughton replied:
i believe it's fixed now. the way it was working was accidental - not what i thought. thanks for the note.

not sure i handled the possibly undefined `configuration.includePath` in a standard way. let me know if you'd like it changed. i appreciate your time on this - i have learned a lot stumbling around with your help.

bmacnaughton committed:
log relevant errors

- remove experimental artifact

sean-mcmanus commented on Extension/src/LanguageServer/configurations.ts:
```diff
@@ -386,6 +395,71 @@ export class CppProperties {
+                for (const dep in nodeAddonMap) {
+                    if (dep in package_json.dependencies) {
+                        const execCmd: string = nodeAddonMap[dep];
+                        let stdout: string = await util.execChildProcess(execCmd, rootPath);
```
This will automatically execute any node or node.exe process in the user's root workspace if they have a package.json, which is potentially unwanted. Could you use something like
```
        const pathToNode: string = which.sync("node");
        const nodeAddonMap: { [dependency: string]: string } = {
            "nan": `"${pathToNode}" --no-warnings -e \"require('nan')\"`,
            "node-addon-api": `"${pathToNode}" --no-warnings -p \"require('node-addon-api').include\"`
        };
```
instead? That will cause the "node" on the user's path to execute instead of potentially one in their current working directory.

bmacnaughton replied:
no problem and good catch. thank you - i had wondered about how to execute "which" cross-platform. that wouldn't execute a local node in linux which is where i spend my time. i hadn't realized that it would do so in windows - a good learning for me.

bmacnaughton committed:
use installed node for requires

bmacnaughton committed:
remove incorrect include concat

sean-mcmanus commented on Extension/src/LanguageServer/configurations.ts:
```diff
@@ -386,6 +391,70 @@ export class CppProperties {
+
+        const pathToNode: string = which.sync("node");
+        const nodeAddonMap: { [dependency: string]: string } = {
+            "nan": `${pathToNode} --no-warnings -e "require('nan')"`,
```
${pathToNode} needs to be `"${pathToNode}"` in order to handle spaces in the path on Windows (I haven't checked Linux).

bmacnaughton replied:
just pushed. it would need them on linux too (though most people don't use spaces in paths). i've been a bit sloppy.

bmacnaughton replied:
would you like me to resolve these when i think i have or would you like to?

sean-mcmanus replied:
Yeah, you can resolve them.

bmacnaughton committed:
quote pathToNode

sean-mcmanus commented on Extension/src/LanguageServer/configurations.ts:
```diff
@@ -627,6 +696,10 @@ export class CppProperties {
             const configuration: Configuration = this.configurationJson.configurations[i];
 
             configuration.includePath = this.updateConfigurationStringArray(configuration.includePath, settings.defaultIncludePath, env);
+            if (settings.addNodeAddonIncludePaths) {
```
configuration.includePath is checked/modified in the code below under certain settings. The "unmodified" configuration.includePath should be saved to some variable here to be used later.

sean-mcmanus replied:
```
if (!configuration.includePath && !!this.defaultIncludes) {
                            configuration.includePath = this.defaultIncludes;
                        }
```
should check the unmodified/original variable and then do something like
```
                const includePath: string[] = configuration.includePath || [];
                configuration.includePath = includePath.concat(this.defaultIncludes);
```

bmacnaughton replied:
thanks for this; i feel like a dog watching tv when i work in this codebase. everything is new and my focus is on a very tiny slice of what is going on.

sean-mcmanus commented on Extension/src/LanguageServer/configurations.ts:
```diff
@@ -288,7 +291,8 @@ export class CppProperties {
     private applyDefaultIncludePathsAndFrameworks(): void {
-        if (this.configurationIncomplete && this.defaultIncludes && this.defaultFrameworks && this.vcpkgPathReady) {
+        if (this.configurationIncomplete && this.defaultIncludes && this.defaultFrameworks && this.vcpkgPathReady
+            && this.nodeAddonIncludePathsReady) {
```
This variable is no longer needed. It was only needed when the applyDefaultConfigurationValues was being used.

bmacnaughton replied:
yes, obvious when i see it. you have a much better handle on all the moving pieces than i do - thanks for spending time on this.

sean-mcmanus commented on Extension/src/LanguageServer/configurations.ts:
```diff
@@ -386,6 +391,70 @@ export class CppProperties {
+                error = e;
+            }
+        }
+        if (error && pdjFound) {
```
I think this should be something like
```
if (error) {
if (pdjFound) {
...
}
} else {
this.handleConfigurationChange();
}
```
i.e. to skip this.handleConfigurationChange when it's not needed.

sean-mcmanus replied:
Or checking this.nodeAddonIncludes.length > 0 (i.e. in case one of the cases errors?).

bmacnaughton replied:
i think it is (conceivably) possible that both node-addon-api and nan are referenced in package.json, both requires succeed, but one of two util.checkDirectoryExists() fail.

but i am changing it as i think it's 1) not possible now given what node-addon-api and nan return and 2) a really tiny edge case going forward because it means the package has to return an invalid directory.

thank you.

sean-mcmanus commented on Extension/src/LanguageServer/configurations.ts:
```diff
@@ -386,6 +391,70 @@ export class CppProperties {
+            .then(pdj => {pdjFound = true; return JSON.parse(pdj); })
+            .catch(e => (error = e));
+
+        const pathToNode: string = which.sync("node");
```
This should be inside the `try {` below (avoids calling which.sync when it's not needed).

bmacnaughton replied:
good point.

sean-mcmanus commented on Extension/src/LanguageServer/client.ts:
```diff
@@ -1367,6 +1367,11 @@ export class DefaultClient implements Client {
+                    // if addNodeAddonIncludePaths was turned on but no includes have been found yet then 1) presume that nan
+                    // or node-addon-api was installed so prompt for reload.
+                    if (changedSettings["addNodeAddonIncludePaths"] && settings.addNodeAddonIncludePaths && this.configuration.nodeAddonIncludesFound() === 0) {
+                        util.promptForReloadWindowDueToSettingsChange();
```
This will only get shown if a user adds the nan/node-addon-api to their package.json and then enables the setting -- is that intentional? It seems fine to me, but it just covers a subset of the cases.

bmacnaughton replied:
> prompts for reload if addNodeAddonIncludePaths is turned on but no addon includes were found by readNodeAddonIncludeLocations. this doesn't catch the case where nan was found but they added node-addon-api before turning on the setting.

above is my original comment on this. it's definitely a subset but seemed like one that is likely to be hit with a typical workflow - create a project, start editing, add node-addon-api or nan, notice the squiggles, and then fix it.

if your experience is different i'm happy to take it out. but it's kind of the best that could be done without creating a file-watcher or changing `readNodeAddonIncludeLocations` to be synchronous.

sean-mcmanus replied:
Yeah, seems fine to me.

bmacnaughton committed:
address review issues

- remove unused flag
- avoid unnecessary work
- handle alternate include setting path
